File: 48-rotate-image/rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        rotate outer layer if the matric and then move in to the next
        """

        l = 0
        r = len(matrix) - 1

        while l < r:
            for i in range(0, r-l):
                t = l
                b = r

                #save topleft
                tl = matrix[t][l+i]

                matrix[t][l+i] = matrix[b-i][l]

                matrix[b-i][l] = matrix[b][r-i]
                
                matrix[b][r-i] = matrix[t+i][r]
                
                matrix[t+i][r] = tl

            r -= 1
            l += 1
        return matrix
            
        

File: 48-rotate-image/rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix[0].size();
        int l = 0;
        int r = n-1;
        int t = 0;
        int b = n-1;

        while (l<r){
            for (int i=0; i<r-l; i++){
                // bl to tl
                int temp = matrix[t][l+i];
                matrix[t][l+i] = matrix[b-i][l];

                // br to bl
                matrix[b-i][l] = matrix[b][r-i];

                //tr to br
                matrix[b][r-i] = matrix[t+i][r];

                
                matrix[t+i][r] = temp;

            }
            l++;
            r--;
            t++;
            b--;
        }
        

    }
};